### ISSUE TITLE
Ucs2904

### ISSUE THREAD

**Issue by m00bh000** [FastLED/FastLED#932]
Please can we get support for ucs2904

**Comment by fishypanther** [comment]
Did this get picked up?

**Comment by psteffensen** [comment]
I am also interested in UCS2094 support. Maybe there is not a big difference between UCS1903 and UCS2904?

**Comment by ewowi** [comment]
+1

**Comment by zackees** [comment]
Not any time soon.

If anyone wants to send me a strip to test then this will accelerate things.

**Comment by zackees** [comment]
Deferred by automated triage: Legitimate open chipset request; maintainer has already stated a position and the blocker is hardware, not effort. zackees (2025-04-12): 'Not any time soon. If anyone wants to send me a strip to test then this will accelerate things.' Deferring rather than closing so the request stays visible against that offer. IMPORTANT CORRECTION recorded on the issue: psteffensen (2024) guessed 'maybe there is not a big difference between UCS1903 and UCS2904'. The in-tree timings show the UCS family is NOT interchangeable -- from src/fl/chipsets/led_timing.h: UCS1903_400KHZ T1=500 T2=1500 T3=500 (T0H 500, T1H 2000, period 2500); UCS1903B_800KHZ T1=400 T2=450 T3=450 (T0H 400, T1H 850, period 1300); UCS1904_800KHZ T1=400 T2=400 T3=450 (T0H 400, T1H 800, period 1250); UCS2903 T1=250 T2=750 T3=250 (T0H 250, T1H 1000, period 1250). UCS2903 vs UCS1904 differ by 150ns on T0H and 200ns on T1H -- far outside typical +/-150ns tolerance. So substituting a neighbouring UCS type is a guess, and this is exactly the failure mode that caused years of intermittent breakage on #739 (GS1903 aliased to WS2812 timing, 50ns out) and #903 (WS2813 driven with WS2812 timing). Unblocking needs the UCS2904 datasheet T0H/T1H/period/reset plus tolerance, or hardware to scope. Anyone with the part can shortcut this: post the four numbers and adding the chipset is a struct in led_timing.h plus a controller alias in chipsets.h. Three requesters over six years (jamesabruce 2019, fishypanther 2022, psteffensen 2024, ewowi 2025).

**Comment by zackees** [comment]
Marking **deferred** — still wanted, blocked on hardware rather than effort, and with one correction that matters.

@zackees stated the position in 2025:

> Not any time soon. If anyone wants to send me a strip to test then this will accelerate things.

Keeping this open and labelled rather than closing, so it stays visible against that offer.

## Correcting an assumption in this thread

@psteffensen wrote:

> Maybe there is not a big difference between UCS1903 and UCS2904? I just found the UCS2903, UCS1903B and UCS1904. All similar to UCS2904.

Reasonable guess, but the UCS family is **not** interchangeable. From `src/fl/chipsets/led_timing.h`:

| chipset | T0H | T1H | bit period |
|---|---|---|---|
| `UCS1903` (400 kHz) | 500 ns | 2000 ns | 2500 ns |
| `UCS1903B` (800 kHz) | 400 ns | 850 ns | 1300 ns |
| `UCS1904` (800 kHz) | 400 ns | 800 ns | 1250 ns |
| `UCS2903` | **250 ns** | **1000 ns** | 1250 ns |

`UCS2903` differs from `UCS1904` by **150 ns on T0H and 200 ns on T1H** — well outside the ±150 ns these parts typically allow. Same-family naming does not imply same timing.

**This exact substitution is what causes long-running intermittent bugs.** Two examples from this tracker, both fixed recently:

- **#739** — `GS1903` was aliased to WS2812 timing (250/875 vs the datasheet's 300/900 at ±50 ns). Six years of "sometimes works, sometimes doesn't".
- **#903** — a user drove WS2813 strips with WS2812 timing after being told the two were identical. They aren't (250/875/280 µs vs 320/960/300 µs), and the wrong answer sent the reporter down a five-year dead end.

So "just try `UCS1904`" is worth an experiment on your bench, but it isn't a substitute for the real numbers, and a strip that *mostly* works is the worst outcome — it looks fine until it doesn't.

## How to unblock this quickly

If you have a UCS2904 strip, you don't need to ship hardware anywhere. Post:

- **T0H / T0L / T1H / T1L** in nanoseconds
- **reset / latch time** (µs)
- the **tolerance** (this is the part that decides whether an existing entry is safe to reuse)
- ideally the datasheet **attached to the issue**, not linked — see #816, where both datasheet links expired and killed the request

With those four numbers this is small: a `TIMING_UCS2904` struct in `led_timing.h` and a controller alias in `chipsets.h`. That's exactly the shape of the #739 fix.

@jamesabruce, @fishypanther, @psteffensen, @ewowi — four people over six years, so the demand is real; it's the specification that's missing.
